File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyCollectionsShouldNotBeEnumerated.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Collections.Specialized;
using System.Linq;
using System.Runtime.Serialization;
using System.Text;

class CollectionTests
{
    private IEnumerable<int> items = new List<int> { 1, 2, 3 };
    private IDictionary<int, int> dictionaryItems = new Dictionary<int, int>();

    private static bool Predicate(int i) => true;
    private static void Action(int i) { }
    private List<int> ListFactory() => new List<int>();
    private HashSet<int> SetFactory() => new HashSet<int>();
    private Queue<int> QueueFactory() => new Queue<int>();
    private Stack<int> StackFactory() => new Stack<int>();
    private ObservableCollection<int> ObservableCollectionFactory() => new ObservableCollection<int>();
    private int[] ArrayFactory() => new int[0];
    private Dictionary<int, int> DictionaryFactory() => new Dictionary<int, int>();

    public void DefaultConstructor()
    {
        var list = new List<int>();
        list.Clear();   // Noncompliant {{Remove this call, the collection is known to be empty here.}}
//      ^^^^^^^^^^^^
        var set = new HashSet<int>();
        set.Clear();    // Noncompliant
        var queue = new Queue<int>();
        queue.Clear();  // Noncompliant
        var stack = new Stack<int>();
        stack.Clear();  // Noncompliant
        var obs = new ObservableCollection<int>();
        obs.Clear();    // Noncompliant
        var array = new int[0];
        array.Clone();  // FIXME Non-compliant
        var dict = new Dictionary<int, int>();
        dict.Clear();   // Noncompliant
    }

    public void ConstructorWithCapacity()
    {
        var list = new List<int>(5);
        list.Clear();   // Noncompliant
        var set = new HashSet<int>(5);
        set.Clear();    // Noncompliant
        var queue = new Queue<int>(5);
        queue.Clear();  // Noncompliant
        var stack = new Stack<int>(5);
        stack.Clear();  // Noncompliant
        var array = new int[5];
        array.Clone();  // Compliant
        var dict = new Dictionary<int, int>(5);
        dict.Clear();   // Noncompliant
    }

    public void ConstructorWithEnumerable()
    {
        var list = new List<int>(items);
        list.Clear();   // Noncompliant FP
        var set = new HashSet<int>(items);
        set.Clear();    // Noncompliant FP
        var queue = new Queue<int>(items);
        queue.Clear();  // Noncompliant FP
        var stack = new Stack<int>(items);
        stack.Clear();  // Noncompliant FP
        var obs = new ObservableCollection<int>(items);
        obs.Clear();    // Noncompliant FP
        var dict = new Dictionary<int, int>(dictionaryItems);
        dict.Clear();   // Noncompliant FP
    }

    public void ConstructorWithEmptyInitializer()
    {
        var list = new List<int> { };
        list.Clear();   // Noncompliant
        var set = new HashSet<int> { };
        set.Clear();    // Noncompliant
        var queue = new Queue<int> { };
        queue.Clear();  // Noncompliant
        var stack = new Stack<int> { };
        stack.Clear();  // Noncompliant
        var obs = new ObservableCollection<int> { };
        obs.Clear();    // Noncompliant
        var array = new int[] { };
        array.Clone();  // FIXME Non-compliant
        var dict = new Dictionary<int, int> { };
        dict.Clear();   // Noncompliant
    }

    public void ConstructorWithInitializer()
    {
        var list = new List<int> { 1, 2, 3 };
        list.Clear();   // Noncompliant FP
        var set = new HashSet<int> { 1, 2, 3 };
        set.Clear();    // Noncompliant FP
        var obs = new ObservableCollection<int> { 1, 2, 3 };
        obs.Clear();    // Noncompliant FP
        var array = new int[] { 1, 2, 3 };
        array.Clone();  // Compliant
        var dict = new Dictionary<int, int>
        {
            [1] = 1,
            [2] = 2,
            [3] = 3,
        };
        dict.Clear();   // Noncompliant FP
    }

    public void Other_Initialization()
    {
        var list = ListFactory();
        list.Clear();   // Compliant
        var set = SetFactory();
        set.Clear();    // Compliant
        var queue = QueueFactory();
        queue.Clear();  // Compliant
        var stack = StackFactory();
        stack.Clear();  // Compliant
        var obs = ObservableCollectionFactory();
        obs.Clear();    // Compliant
        var array = ArrayFactory();
        array.Clone();  // Compliant
        var dict = DictionaryFactory();
        dict.Clear();   // Compliant
    }

    public void Methods_Raise_Issue()
    {
        int i;

        var list = new List<int>();
        list.BinarySearch(5);           // Noncompliant
        list.Clear();                   // Noncompliant
        list.Contains(5);               // Noncompliant
        list.ConvertAll(x => x);        // Noncompliant
        list.CopyTo(null, 1);           // Noncompliant
        list.Exists(Predicate);         // Noncompliant
        list.Find(Predicate);           // Noncompliant
        list.FindAll(Predicate);        // Noncompliant
        list.FindIndex(Predicate);      // Noncompliant
        list.FindLast(Predicate);       // Noncompliant
        list.FindLastIndex(Predicate);  // Noncompliant
        list.ForEach(Action);           // Noncompliant
        list.GetEnumerator();           // Noncompliant
        list.GetRange(1, 5);            // Noncompliant
        list.IndexOf(5);                // Noncompliant
        list.LastIndexOf(5);            // Noncompliant
        list.Remove(5);                 // Noncompliant
        list.RemoveAll(Predicate);      // Noncompliant
        list.RemoveAt(1);               // Noncompliant
        list.RemoveRange(1, 5);         // Noncompliant
        list.Reverse();                 // Noncompliant
        list.Sort();                    // Noncompliant
        list.TrueForAll(Predicate);     // Noncompliant
        _ = list[1];                    // FIXME Non-compliant
//          ~~~~~~~
        list[1] = 5;                    // FIXME Non-compliant
//      ~~~~~~~

        var set = new HashSet<int>();
        set.Clear();                    // Noncompliant
        set.Contains(5);                // Noncompliant
        set.CopyTo(null, 1);            // Noncompliant
        set.ExceptWith(items);          // Noncompliant
        set.GetEnumerator();            // Noncompliant
        set.IntersectWith(items);       // Noncompliant
        set.IsProperSubsetOf(items);    // Noncompliant
        set.IsProperSupersetOf(items);  // Noncompliant
        set.IsSubsetOf(items);          // Noncompliant
        set.IsSupersetOf(items);        // Noncompliant
        set.Overlaps(items);            // Noncompliant
        set.Remove(5);                  // Noncompliant
        set.RemoveWhere(Predicate);     // Noncompliant
        set.SymmetricExceptWith(items); // Noncompliant, also learns NotEmpty
        set = new HashSet<int>();
        set.TryGetValue(5, out i);      // Noncompliant
        set.UnionWith(items);           // Noncompliant, also learns NotEmpty

        var queue = new Queue<int>();
        queue.Clear();                  // Noncompliant
        queue.Contains(5);              // Noncompliant
        queue.CopyTo(null, 1);          // Noncompliant
        queue.Dequeue();                // Noncompliant
        queue.GetEnumerator();          // Noncompliant
        queue.Peek();                   // Noncompliant

        var stack = new Stack<int>();
        stack.Clear();                  // Noncompliant
        stack.Contains(5);              // Noncompliant
        stack.CopyTo(null, 0);          // Noncompliant
        stack.GetEnumerator();          // Noncompliant
        stack.Peek();                   // Noncompliant
        stack.Pop();                    // Noncompliant

        var obs = new ObservableCollection<int>();
        obs.Clear();                    // Noncompliant
        obs.Contains(5);                // Noncompliant
        obs.CopyTo(null, 1);            // Noncompliant
        obs.GetEnumerator();            // Noncompliant
        obs.IndexOf(5);                 // Noncompliant
        obs.Move(1, 2);                 // Noncompliant
        obs.Remove(5);                  // Noncompliant
        obs.RemoveAt(1);                // Noncompliant
        _ = obs[1];                     // FIXME Non-compliant
        obs[1] = 5;                     // FIXME Non-compliant

        var array = new int[0];
        array.Clone();                  // FIXME Non-compliant
        array.CopyTo(null, 1);          // FIXME Non-compliant
        array.GetEnumerator();          // FIXME Non-compliant
        array.GetLength(1);             // FIXME Non-compliant
        array.GetLongLength(1);         // FIXME Non-compliant
        array.GetLowerBound(1);         // FIXME Non-compliant
        array.GetUpperBound(1);         // FIXME Non-compliant
        array.GetValue(1);              // FIXME Non-compliant
        array.Initialize();             // FIXME Non-compliant
        array.SetValue(5, 1);           // FIXME Non-compliant
        _ = array[1];                   // FIXME Non-compliant
        array[1] = 5;                   // FIXME Non-compliant

        var dict = new Dictionary<int, int>();
        dict.Clear();                   // Noncompliant
        dict.ContainsKey(1);            // Noncompliant
        dict.ContainsValue(5);          // Noncompliant
        dict.GetEnumerator();           // Noncompliant
        dict.Remove(5);                 // Noncompliant
        dict.TryGetValue(1, out i);     // Noncompliant
        _ = dict[1];                    // FIXME Non-compliant
    }

    public void Methods_Ignored()
    {
        int i;

        var list = new List<int>();
        list.AsReadOnly();
        list.GetHashCode();
        list.GetType();
        list.Equals(items);
        list.ToString();
        list.TrimExcess();
        list.ToArray();

        var set = new HashSet<int>();
        set.Equals(items);
        set.GetHashCode();
        set.GetObjectData(null, new StreamingContext());
        set.GetType();
        set.OnDeserialization(null);
        set.SetEquals(items);
        set.TrimExcess();

        var queue = new Queue<int>();
        queue.Equals(items);
        queue.GetHashCode();
        queue.GetType();
        queue.ToArray();
        queue.ToString();
        queue.TrimExcess();

        var stack = new Stack<int>();
        stack.GetHashCode();
        stack.Equals(items);
        stack.GetType();
        stack.ToArray();
        stack.ToString();
        stack.TrimExcess();

        var obs = new ObservableCollection<int>();
        obs.GetHashCode();
        obs.Equals(items);
        obs.GetType();
        obs.ToString();
        _ = obs.Count;
        obs.CollectionChanged += (s, e) => throw new NotImplementedException();

        var array = new int[0];
        array.GetHashCode();
        array.Equals(new object());
        array.GetType();
        array.ToString();
        _ = array.Length;

        var dict = new Dictionary<int, int>();
        dict.GetHashCode();
        dict.GetObjectData(null, new StreamingContext());
        dict.Equals(items);
        dict.GetType();
        dict.OnDeserialization(null);
        dict.ToString();
        dict[5] = 5;
        (((dict[5]))) = 5;
    }

    public void Methods_Set_NotEmpty()
    {
        var list = new List<int>();
        list.Add(5);
        list.Clear();                   // Compliant
        list = new List<int>();
        list.AddRange(items);
        list.Clear();                   // Compliant
        list = new List<int>();
        list.Insert(1, 5);
        list.Clear();                   // Compliant
        list = new List<int>();
        list.InsertRange(1, items);
        list.Clear();                   // Compliant

        var set = new HashSet<int>();
        set.Add(1);
        set.Clear();                    // Compliant
        set = new HashSet<int>();
        set.SymmetricExceptWith(items); // Noncompliant
        set.Clear();                    // Compliant
        set = new HashSet<int>();
        set.UnionWith(items);           // Noncompliant
        set.Clear();                    // Compliant

        var queue = new Queue<int>();
        queue.Enqueue(5);
        queue.Clear();                  // Compliant

        var stack = new Stack<int>();
        stack.Push(5);
        stack.Clear();                  // Compliant

        var obs = new ObservableCollection<int>();
        obs.Add(5);
        obs.Clear();                    // Compliant
        obs = new ObservableCollection<int>();
        obs.Insert(0, 5);
        obs.Clear();                    // Compliant

        var dict = new Dictionary<int, int>();
        dict.Add(1, 5);
        dict.Clear();                   // Compliant
    }
}

class AdvancedTests
{
    public void ExtensionMethods_Should_Not_Raise()
    {
        var list = new List<int>();
        list.Any();     // Compliant
    }

    public void ExtensionMethods_Remove_Constraints()
    {
        var list = new List<int>();
        list.Any();
        list.Clear();   // Noncompliant FP
    }

    public void PassingAsArgument_Removes_Constraints()
    {
        var list = new List<int>();
        Foo(list);
        list.Clear();   // Noncompliant FP
    }

    public void HigherRank_And_Jagged_Array_Is_NotEmpty()
    {
        var array1 = new int[0, 0];
        array1.Clone(); // Compliant
        int[][] array2 = new int[1][];
        array2.Clone(); // Compliant
    }

    void Foo(IEnumerable<int> items) { }
}

// This simulates the Dictionary from .NetCore 2.0+.
public class NetCoreDictionary<TKey, TValue> : Dictionary<TKey, TValue>
{
    public void TestTryAdd()
    {
        var dict = new NetCoreDictionary<string, object>();
        if (dict.TryAdd("foo", new object()))   // Compliant
        {
        }
    }

    public bool TryAdd(TKey key, TValue value)
    {
        return true;
    }
}

class Flows
{
    public void Conditional_Add(bool condition)
    {
        var list = new List<int>();
        if (condition)
        {
            list.Add(5);
        }
        list.Clear();   // Compliant
    }

    public void Conditional_Add_With_Loop(bool condition)
    {
        var list = new List<int>();
        while (true)
        {
            if (condition)
            {
                list.Add(5);
                break;
            }
        }
        list.Clear();   // Compliant
    }

    public void AddPassedAsParameter()
    {
        var list = new List<int>();

        DoSomething(list.Add);

        list.Clear();   // Noncompliant FP, see https://github.com/SonarSource/sonar-dotnet/issues/4261
    }

    private static void DoSomething(Action<int> callback) => callback(42);
}


class Flows2
{
    public static string UrlDecode(string s, Encoding e)
    {
        long len = s.Length;
        var bytes = new List<byte>();
        int xchar;
        char ch;

        for (int i = 0; i < len; i++)
        {
            ch = s[i];
            if (ch == '%' && i + 2 < len && s[i + 1] != '%')
            {
                if (s[i + 1] == 'u' && i + 5 < len)
                {
                    // unicode hex sequence
                    xchar = GetChar(s, i + 2, 4);
                    if (xchar != -1)
                    {
                        WriteCharBytes(bytes, (char)xchar, e);
                        i += 5;
                    }
                    else
                        WriteCharBytes(bytes, '%', e);
                }
                else if ((xchar = GetChar(s, i + 1, 2)) != -1)
                {
                    WriteCharBytes(bytes, (char)xchar, e);
                    i += 2;
                }
                else
                {
                    WriteCharBytes(bytes, '%', e);
                }
                continue;
            }

            if (ch == '+')
                WriteCharBytes(bytes, ' ', e);
            else
                WriteCharBytes(bytes, ch, e);
        }

        byte[] buf = bytes.ToArray();
        bytes = null;
        return e.GetString(buf);

    }

    private static void WriteCharBytes(List<byte> bytes, char v, Encoding e)
    {
    }

    private static int GetChar(string s, int v1, int v2)
    {
        return 1;
    }
}

// See https://github.com/SonarSource/sonar-dotnet/issues/1002
class Program
{
    public Program(string id)
    {
        if (string.IsNullOrEmpty(id))
            throw new ArgumentNullException(nameof(id));
    }
}

class LargeCfg
{
    // Large CFG that causes the exploded graph to hit the exploration limit
    // See https://github.com/SonarSource/sonar-dotnet/issues/767 (comments!)
    private static void MethodName(object[] table, object[][] aoTable2, ref Dictionary<double, string> dictPorts)
    {
        try
        {
            string sValue = "-1";
            switch (5)
            {
                case 3:
                    sValue = "";
                    break;
                case 4:
                    sValue = "";
                    break;
                case 5:
                    sValue = "";
                    break;
            }

            var list = new List<double>();
            for (int i = 0; i < table.Length; i++)
            {
                if (Convert.ToInt32(table[3]) == 1 /* Normal */ &&
                    Convert.ToInt32(aoTable2[5][9]) == 1 /* On */)
                {
                    string sValue2;
                    switch (Convert.ToInt32(aoTable2[5][4]))
                    {
                        default:
                            sValue2 = "";
                            break;
                    }

                    if (dictPorts.ContainsKey(5))
                    {
                        list.Add(7);
                    }
                }
                else
                {
                    dictPorts.Add(5, sValue);
                }
            }

            if (list.Count > 0)
            {
                list.Sort();    // Compliant - used to be an FP
            }
        }
        catch
        {
            // silently do nothing
        }
    }
}

// https://github.com/SonarSource/sonar-dotnet/issues/4478
public class Repro_4478
{
    public string Main()
    {
        var list = new List<String>();
        AddInLocalFunction();
        return list[0]; // FIXME Non-compliant FP

        void AddInLocalFunction()
        {
            list.Add("Item1");
        }
    }
}

File: analyzers/tests/SonarAnalyzer.Test/TestCases/SymbolicExecution/Roslyn/EmptyCollectionsShouldNotBeEnumerated.vb
Imports System.Collections.Generic
Imports System.Collections.ObjectModel

Public Class CollectionTests

    Private Items As IEnumerable(Of Integer)
    Private Predicate As Predicate(Of Integer)
    Private Action As Action(Of Integer)

    Private Function GetList() As List(Of Integer)
    End Function

    Public Sub DefaultConstructor()
        Dim List As New List(Of Integer)
        List.Clear()                    ' Noncompliant {{Remove this call, the collection is known to be empty here.}}

        Dim HS As New HashSet(Of Integer)
        HS.Clear()                      ' Noncompliant
        Dim Q As New Queue(Of Integer)
        Q.Clear()                       ' Noncompliant
        Dim S As New Stack(Of Integer)
        S.Clear()                       ' Noncompliant
        Dim OC As New ObservableCollection(Of Integer)
        OC.Clear()                      ' Noncompliant
        Dim Arr(-1) As Integer
        Arr.Clone()                     ' Noncompliant
        Dim D As New Dictionary(Of Integer, Integer)
        D.Clear()                       ' Noncompliant
    End Sub

    Public Sub ConstructorWithCapacity()
        Const ConstCount0 As Integer = -1
        Const ConstCount5 As Integer = 5
        Dim VarCount0 As Integer = -1
        Dim VarCount5 As Integer = 5

        Dim List As New List(Of Integer)(5)
        List.Clear()                        ' Noncompliant
        Dim ArrLiteral5(5) As Integer
        ArrLiteral5.Clone()                 ' Compliant
        Dim ArrLiteral0(-1) As Integer
        ArrLiteral0.Clone()                 ' Noncompliant
        Dim ArrConst5(ConstCount5) As Integer
        ArrConst5.Clone()                   ' Compliant
        Dim ArrConst0(ConstCount0) As Integer
        ArrConst0.Clone()                   ' Noncompliant
        Dim ArrVar5(VarCount5) As Integer
        ArrConst5.Clone()                   ' Compliant
        Dim ArrVar0(VarCount0) As Integer
        ArrConst0.Clone()                   ' Noncompliant

        Dim ArrBinary(VarCount5 - ConstCount5 - 1) As Integer
        ArrBinary.Clone()                   ' FN
    End Sub

    Public Sub ConstructorWithEnumerable()
        Dim List As New List(Of Integer)(Items)
        List.Clear()                    ' Compliant
    End Sub

    Public Sub ConstructorWithEnumerableWithConstraint(Condition As Boolean)
        Dim BaseCollection As New List(Of Integer)
        Dim HS As New HashSet(Of Integer)(BaseCollection)
        HS.Clear()                      ' Noncompliant

        BaseCollection = New List(Of Integer)
        HS = New HashSet(Of Integer)(BaseCollection, EqualityComparer(Of Integer).Default)
        HS.Clear()                      ' Noncompliant

        BaseCollection = New List(Of Integer)
        HS = New HashSet(Of Integer)(comparer:=EqualityComparer(Of Integer).Default, collection:=BaseCollection)
        HS.Clear()                      ' Noncompliant

        BaseCollection = New List(Of Integer)
        HS = New HashSet(Of Integer)(If(Condition, BaseCollection, BaseCollection))
        HS.Clear()                      ' Noncompliant

        BaseCollection.Add(1)
        HS = New HashSet(Of Integer)(BaseCollection)
        HS.Clear()                      ' Compliant
    End Sub

    Public Sub ConstructorWithEmptyInitializer()
        Dim List As New List(Of Integer) From {}
        List.Clear()                    ' Noncompliant
    End Sub

    Public Sub ConstructorWithInitializer()
        Dim List As New List(Of Integer) From {1, 2, 3}
        List.Clear()                    ' Compliant
        Dim Array() As Integer = {1, 2, 3}
        Array.Clone()                   ' Compliant
    End Sub

    Public Sub Other_Initialization()
        Dim List As List(Of Integer) = GetList()
        List.Clear()                    ' Compliant
        Dim Array As Integer() = Array.Empty(Of Integer)
        Array.Clone()                   ' FN
        Dim Enm As IEnumerable(Of Integer) = Enumerable.Empty(Of Integer)
        Enm.GetEnumerator()             ' FN
    End Sub

    Public Sub Methods_Raise_Issue()
        Dim List As New List(Of Integer), i As Integer
        List.BinarySearch(5)            ' Noncompliant
        List.Clear()                    ' Noncompliant
        List.Contains(5)                ' Noncompliant
        List.ConvertAll(Function(X) X)  ' Noncompliant
        List.CopyTo(Nothing, 1)         ' Noncompliant
        List.Exists(Predicate)          ' Noncompliant
        List.Find(Predicate)            ' Noncompliant
        List.FindAll(Predicate)         ' Noncompliant
        List.FindIndex(Predicate)       ' Noncompliant
        List.FindLast(Predicate)        ' Noncompliant
        List.FindLastIndex(Predicate)   ' Noncompliant
        List.ForEach(Action)            ' Noncompliant
        List.GetEnumerator()            ' Noncompliant
        List.GetRange(1, 5)             ' Noncompliant
        List.IndexOf(5)                 ' Noncompliant
        List.LastIndexOf(5)             ' Noncompliant
        List.Remove(5)                  ' Noncompliant
        List.RemoveAll(Predicate)       ' Noncompliant
        List.RemoveAt(1)                ' Noncompliant
        List.RemoveRange(1, 5)          ' Noncompliant
        List.Reverse()                  ' Noncompliant
        List.Sort()                     ' Noncompliant
        List.TrueForAll(Predicate)      ' Noncompliant
        i = List(1)                     ' Compliant, should be part of S6466
        List(1) = 5                     ' Compliant, should be part of S6466
        Dim ToArr() As Integer = List.ToArray
        ToArr.Clone()                   ' FN

        Dim Array(-1) As Integer
        Array.Clone()                   ' Noncompliant
        Array.CopyTo(Nothing, 1)        ' Noncompliant
        Array.GetEnumerator()           ' Noncompliant
        Array.GetLength(1)              ' Noncompliant
        Array.GetLongLength(1)          ' Noncompliant
        Array.GetLowerBound(1)          ' Noncompliant
        Array.GetUpperBound(1)          ' Noncompliant
        Array.GetValue(1)               ' Noncompliant
        Array.Initialize()              ' Noncompliant
        Array.SetValue(5, 1)            ' Noncompliant
        i = Array(1)                    ' Compliant, should be part of S6466
        Array(1) = 5                    ' Compliant, should be part of S6466
    End Sub

    Public Sub Methods_Ignored()
        Dim List As New List(Of Integer)
        List.AsReadOnly()
        List.GetHashCode()
        List.GetType()
        List.Equals(Items)
        List.ToString()
        List.TrimExcess()
        List.ToArray()

        Dim Array(5) As Integer
        Array.GetHashCode()
        Array.Equals(New Object())
        Array.GetType()
        Array.ToString()
        Dim Length As Integer = Array.Length
    End Sub

    Public Sub Methods_Set_NotEmpty()
        Dim List As New List(Of Integer)
        List.Add(5)
        List.Clear()                   ' Compliant
        List = New List(Of Integer)
        List.AddRange(Items)
        List.Clear()                   ' Compliant
        List = New List(Of Integer)
        List.Insert(1, 5)
        List.Clear()                   ' Compliant
        List = New List(Of Integer)
        List.InsertRange(1, Items)
        List.Clear()                   ' Compliant
    End Sub

    Public Sub Method_Set_Empty(List As List(Of Integer))
        List.Clear()                    ' Compliant
        List.Clear()                    ' Noncompliant

        Dim Empty As New List(Of Integer)
        List.Add(5)
        List.Intersect(Empty)           ' Compliant
        List.Clear()                    ' FN
    End Sub

    Public Sub ForEach()
        Dim List As New List(Of Integer)
        For Each i As Integer In List   ' Noncompliant
        Next
    End Sub

    Public Sub Casing()
        Dim List As New List(Of Integer)
        List.Clear()                    ' Noncompliant
    End Sub

    Public Sub LearnConditions_Size(Condition As Boolean)
        Dim IsNull As List(Of Integer) = Nothing
        Dim Empty As New List(Of Integer)
        Dim NotEmpty As New List(Of Integer) From {1, 2, 3}

        If If(IsNull, Empty).Count = 0 Then Empty.Clear()  ' Noncompliant
        If If(IsNull, NotEmpty).Count = 0 Then Empty.Clear()  ' Compliant, unreachable
        If If(Condition, Empty, Empty).Count = 0 Then
            Empty.Clear()  ' Noncompliant
        Else
            Empty.Clear()  ' Compliant, unreachable
        End If
        If Empty.Count = 0 Then
            Empty.Clear()  ' Noncompliant
        Else
            Empty.Clear()  ' Compliant, unreachable
        End If
        If Empty.Count() = 0 Then
            Empty.Clear()  ' Noncompliant
        Else
            Empty.Clear()  ' Compliant, unreachable
        End If
        If DirectCast(Empty, IEnumerable(Of Integer)).Count(Function(X) Condition) = 0 Then
            Empty.Clear()   ' Noncompliant
        Else
            Empty.Clear()   ' Compliant, unreachable
        End If
        If DirectCast(NotEmpty, IEnumerable(Of Integer)).Count(Function(X) Condition) = 0 Then
            Empty.Clear()   ' Noncompliant
        Else
            Empty.Clear()   ' Noncompliant
        End If
        If Enumerable.Count(Empty) = 0 Then
            Empty.Clear()  ' Noncompliant
        Else
            Empty.Clear()  ' Compliant, unreachable
        End If
        NotEmpty.Clear()   ' Compliant, prevents LVA from throwing notEmpty away during reference capture
    End Sub

    Public Sub Enumerable_ExtensionIsInstance()
        Dim Empty As New List(Of Integer)
        Dim AsIEnumerable As IEnumerable(Of Integer) = Empty
        If AsIEnumerable.Count = 0 Then ' This is invoked as instance, not as a static call
            Empty.Clear()   ' Noncompliant
        Else
            Empty.Clear()   ' Compliant, unreachable
        End If
    End Sub

End Class

Public Class AdvancedTests

    Public Sub UnknownExtensionMethods()
        Dim List As New List(Of Integer)
        List.CustomExtensionMethod()    ' Compliant
        List.Clear()                    ' Noncompliant FP
    End Sub

    Public Sub WellKnownExtensionMethods()
        Dim List As New List(Of Integer)
        List.All(Function(X) True)      ' FN
        List.Any()                      ' FN
        Enumerable.Reverse(List)        ' Noncompliant
        List.Clear()                    ' Noncompliant
    End Sub

    Public Sub PassingAsArgument_Removes_Constraints()
        Dim List As New List(Of Integer)
        Foo(List)
        List.Clear()   ' Compliant
    End Sub

    Public Sub HigherRank_And_Jagged_Array()
        Dim Array1(-1, -1) As Integer
        Array1.Clone()                  ' Noncompliant
        Dim Array2(-1, 4) As Integer
        Array2.Clone()                  ' Noncompliant
        Dim Array3(5, 4) As Integer
        Array3.Clone()                  ' Compliant
        Dim Array4()() As Integer = New Integer(-1)() {}
        Array4.Clone()                  ' Noncompliant
        Dim Array5()() As Integer = New Integer(1)() {}
        Array5.Clone()                  ' Compliant
    End Sub

    Private Sub Foo(Items As IEnumerable(Of Integer))
    End Sub

End Class

Public Module CustomExtensions

    <Runtime.CompilerServices.Extension>
    Public Sub CustomExtensionMethod(List As List(Of Integer))
    End Sub

End Module

' This simulates the Dictionary from .NetCore 2.0+.
Public Class NetCoreDictionary(Of TKey, TValue)
    Inherits Dictionary(Of TKey, TValue)

    Public Sub TestTryAdd()
        Dim Dict As New NetCoreDictionary(Of String, Object)
        If Dict.TryAdd("foo", New Object()) Then   ' Compliant
        End If
    End Sub

    Public Function TryAdd(Key As TKey, Value As TValue) As Boolean
    End Function

End Class

Public Class Flows

    Public Sub Conditional_Add(Condition As Boolean)
        Dim List As New List(Of Integer)
        If Condition Then List.Add(5)
        List.Clear()   ' Compliant
    End Sub

    Public Sub Conditional_Add_With_Loop(Condition As Boolean)
        Dim List As New List(Of Integer)
        While True
            If Condition Then List.Add(5)
            Exit While
        End While
        List.Clear()   ' Compliant
    End Sub

    ' https://github.com/SonarSource/sonar-dotnet/issues/4261
    Public Sub AddPassedAsParameter()
        Dim List As New List(Of Integer)
        DoSomething(AddressOf List.Add)
        List.Clear()   ' Compliant
    End Sub

    Public Sub CountZero(List As List(Of Integer))
        If List.Count = 0 Then
            List.Clear()        ' Noncompliant
        Else
            List.Clear()        ' Compliant
        End If
    End Sub

    Public Sub CountZeroExtension(List As List(Of Integer))
        If Enumerable.Count(List) = 0 Then
            List.Clear()        ' FN
        Else
            List.Clear()        ' Compliant
        End If
    End Sub

    Public Sub CountNotZero(List As List(Of Integer))
        If List.Count <> 0 Then
            List.Clear()        ' Compliant
        Else
            List.Clear()        ' Noncompliant
        End If
    End Sub

    Public Sub CountGreaterThanZero(List As List(Of Integer))
        If List.Count > 0 Then
            List.Clear()        ' Compliant
        Else
            List.Clear()        ' Noncompliant
        End If
    End Sub

    Public Sub CountGreaterThanOne(List As List(Of Integer))
        If List.Count > 1 Then
            List.Clear()        ' Compliant
        Else
            List.Clear()        ' Compliant
        End If
    End Sub

    Private Shared Sub DoSomething(Callback As Action(Of Integer))
    End Sub

End Class
